### PR TITLE
Add changelog generation

### PR DESCRIPTION
**Pull request by markusweigelt**
- only the PR`s analogous to the changelog generation of GitHub are used
- Identifier in merges are used to retrieve the PR information from GitHub 
- currently the following adjustments are still necessary manually:
  - change of the release version number
  - change of the from and end commit references
- manual execution if `update_changelog.sh` using personal token as parameter

### CODE REVIEW

File: update_changelog.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            repo="slub/ocrd_manager"
+        elif [[ $repo == "slub/ocrd_manager" && $pull_request_id == "1" ]]; then
+            repo="slub/ocrd_monitor"
+        fi
+
+        url="https://github.com/$repo"
+
+        #echo "curl -L -H Accept: application/vnd.github+json -H Authorization: Bearer $token  https://api.github.com/repos/$repo/pulls/$pull_request_id"
+        curl -L -H "Accept: application/vnd.github+json" -H "Authorization: Bearer $token"  https://api.github.com/repos/$repo/pulls/$pull_request_id | \
+        jq -r '[.title, .user.login] | @sh' | xargs printf "* %s by @%s in $url/pull/$pull_request_id\n"
```
Comment (bertsky) [inline update_changelog.sh:38]: nice pattern!

File: update_changelog.sh [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+generate_changelog () {
+    local path=$1
+    local title="$2"
+    local fromCommit="$3"
+    local toCommit="$4"
+
+    echo "### $title"
+    echo ""
+    git -C $path log  --pretty="%s" --grep="Merge pull request #" $fromCommit.. | while read -r line ; do
```
Comment (bertsky) [inline update_changelog.sh:21]: This assumes we never alter(ed) the default commit description when merging on Github. But I see no other option. Using `--merges` would also give us updates (merges of master into a feature branch).

But is there a reason you don't use the upper limit of the commit range here?
```suggestion
    git -C $path log  --pretty="%s" --grep="Merge pull request #" $fromCommit..$toCommit | while read -r line ; do
```

File: update_changelog.sh [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        arr=($line)
+        pull_request_id="${arr[3]:1}"
+        repo="$5"
+
+        # deprecated - support old prs of OCR-D Monitor because of moving to separate repository
+        if [[ $repo == "slub/ocrd_monitor" && ( $pull_request_id == "45" || $pull_request_id == "44" || $pull_request_id == "37" || $pull_request_id == "36" ) ]]; then
+            repo="slub/ocrd_manager"
+        elif [[ $repo == "slub/ocrd_manager" && $pull_request_id == "1" ]]; then
+            repo="slub/ocrd_monitor"
+        fi
```
Comment (bertsky) [inline update_changelog.sh:32]: ok, I see the problem and like the pragmatic solution!

**Review by bertsky — APPROVED** [review]
Looks good, thanks!

> * only the PR`s analogous to the changelog generation of GitHub are used

Yes, I guess since we don't do releases or semantic versioning anywhere, this is as good as it gets. PR titles are short but very descriptive – and the list of changes is already long (at least for this release)...

> * Identifier in merges are used to retrieve the PR information from GitHub

There's probably a way to get to that directly via Github API, but let's stick with your method as long as it works.

> * currently the following adjustments are still necessary manually:
>   
>   * change of the release version number
>   * change of the from and end commit references

I would expect your script to be embedded in a release procedure where a new tag is available on the HEAD, and so you would just need to go back to the previous tag to get a commit range. Or, for manual override, optional parameters.

> * manual execution if `update_changelog.sh` using personal token as parameter

And from CI this could be passed on from Github Actions perhaps?

**Comment by markusweigelt** [comment]
- Update script to retrieve commit references without manual intervention
- Improve script to add release changes at the top of existing `CHANGELOG.md`
- Move `update_changelog.sh` to  `.github/scripts`
- Add workflow to provide a PR with changes for release using the `update_changelog.sh` script

**Review by bertsky — APPROVED** [review]
Fantastic!

**Pull request merged**